edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/device/DeviceMessageHandler.cs

```diff
@@ -334,11 +345,53 @@ public static void ProcessedGetTwin(string identityId)
             {
                 Log.LogDebug((int)EventIds.ProcessedGetTwin, Invariant($"Processed GetTwin for {identityId}"));
             }
+
+            public static void C2dNoLockToken(IIdentity identity)
+            {
+                Log.LogWarning((int)EventIds.C2dNoLockToken, Invariant($"Received C2D message for {identity.Id} with no lock token. Abandoning message."));
+            }
+
+            public static void SendingC2DMessage(IIdentity identity, string lockToken)
+            {
+                Log.LogDebug((int)EventIds.SendingC2DMessage, Invariant($"Sending C2D message with lock token {lockToken} to {identity.Id}"));
+            }
+
+            public static void ReceivedC2DMessageWithSameToken(IIdentity identity, string lockToken)
+            {
+                Log.LogWarning((int)EventIds.ReceivedC2DMessageWithSameToken, Invariant($"Received duplicate C2D message for {identity.Id} with the same lock token {lockToken}. Abandoning message."));
+            }
+
+            public static void ReceivedC2DFeedbackMessage(IIdentity identity, string messageId)
+            {
+                Log.LogDebug((int)EventIds.ReceivedC2DFeedbackMessage, Invariant($"Received C2D feedback message from {identity.Id} with lock token {messageId}"));
+            }
+
+            public static void UnknownFeedbackMessage(IIdentity identity, string messageId, FeedbackStatus feedbackStatus)
+            {
+                Log.LogWarning((int)EventIds.UnknownFeedbackMessage, Invariant($"Received unknown feedback message from {identity.Id} with lock token {messageId} and status {feedbackStatus}. Abandoning message."));
+            }
         }
 
         #region IDeviceProxy
 
-        public Task SendC2DMessageAsync(IMessage message) => this.underlyingProxy.SendC2DMessageAsync(message);
+        public Task SendC2DMessageAsync(IMessage message)
+        {
+            if (!message.SystemProperties.TryGetValue(SystemProperties.LockToken, out string lockToken) || string.IsNullOrWhiteSpace(lockToken))
+            {
+                // TODO - Should we throw here instead?
```

Comment:
Is this TODO is something we should resolve before we merge this PR?

Reply:
Not necessarily.. I am not sure of the right behavior here. So added a comment. 
If you or someone else has any feedback on this, I would like to hear it..

Reply:
OK, just double-checking.